packages/lab/src/stepped-tracker/utils.ts

```diff
@@ -6,17 +6,39 @@ export function assignSteps(
   stage?: StepStage,
 ): StepRecord[] {
   return steps.map((step) => {
-    step.stage = stage;
-    if (step.substeps) {
-      step.substeps = assignSteps(step.substeps, stage);
+    if (!step.substeps) {
+      return { ...step, stage };
     }
 
-    return step;
+    return {
+      ...step,
+      stage,
+      substeps: assignSteps(step.substeps, stage),
+    };
   });
 }
 
-export function resetSteps(steps: StepRecord[]): StepRecord[] {
-  return assignSteps(steps, undefined);
+export function resetSteps(
```

Comment:
Would this work ?

```
export function resetSteps(
  steps: StepRecord[],
  { resetStatus = false } = {}
): StepRecord[] {
  return steps.map((step) => ({
    ...step,
    stage: undefined,
    status: resetStatus ? undefined : step.status,
    substeps: step.substeps ? resetSteps(step.substeps, { resetStatus }) : undefined,
  }));
}
```

Reply:
Both achieve the exact same thing, I just like having a clearly defined exit condition at the top of a recursive function, makes it a touch more readable.